stage1/prepare-app/prepare-app.c

```diff
@@ -125,44 +126,78 @@ static int ensure_etc_hosts_exists(const char *root, int rootfd) {
 	return 0;
 }
 
+static void mount_at(const char *root, const mount_point *mnt)
+{
+	char to[4096];
+	exit_if(snprintf(to, sizeof(to), "%s/%s", root, mnt->target) >= sizeof(to),
+		"Path too long: \"%s\"", to);
+	pexit_if(mount(mnt->source, to, mnt->type,
+		       mnt->flags, mnt->options) == -1,
+		 "Mounting \"%s\" on \"%s\" failed", mnt->source, to);
+}
+
 static void mount_sys(const char *root)
 {
-	char from[4096];
+	int i;
 	char to[4096];
 	struct statfs fs;
 	DIR *dir = NULL;
 	struct dirent *d;
+	const mount_point mnt_rec = { "/sys", "sys", "bind", NULL, MS_BIND|MS_REC };
+	const mount_point sys_bind_table[] = {
+		{ "/sys", "sys", "bind", NULL, MS_BIND },
+		{ "/sys/fs/cgroup", "sys/fs/cgroup", "bind", NULL, MS_BIND },
+	};
 
 	pexit_if(statfs("/sys/fs/cgroup", &fs) != 0,
 	         "Cannot statfs /sys/fs/cgroup");
 	if (fs.f_type == (typeof(fs.f_type)) CGROUP2_SUPER_MAGIC) {
 		/* With the unified cgroup hierarchy, recursive bind mounts
 		 * are fine. */
-		exit_if(snprintf(to, sizeof(to), "%s/%s", root, "sys") >= sizeof(to),
-			"Path too long: \"%s\"", to);
-		pexit_if(mount("/sys", to, "bind",
-			       MS_BIND | MS_REC, "NULL") == -1,
-				"Mounting \"%s\" on \"%s\" failed", "/sys", to);
+		mount_at(root, &mnt_rec);
 		return;
 	}
 
+	// For security reasons recent Linux kernels do not allow to bind-mount non-recursively
+	// if it would give read-write access to other subdirectories mounted as read-only.
+	// Hence we have to check if we are in a user namespaced environment and bind mount recursively instead.
+	if (access("/proc/1/uid_map", F_OK) == 0) {
+		FILE *f;
+		int k;
+		uid_t uid_base, uid_shift, uid_range;
+
+		pexit_if((f = fopen("/proc/1/uid_map", "re")) == NULL,
+			 "Unable to open /proc/1/uid_map");
+
+		if (sizeof(uid_t) == 4) {
+			k = fscanf(f, "%"PRIu32" %"PRIu32" %"PRIu32,
+				   &uid_base, &uid_shift, &uid_range);
+		} else {
+			k = fscanf(f, "%"PRIu16" %"PRIu16" %"PRIu16,
+				   &uid_base, &uid_shift, &uid_range);
+		}
+		pexit_if(fclose(f) != 0, "Unable to close /proc/1/uid_map");
+		pexit_if(k != 3, "Invalid uid_map format");
+
+		// do a recursive bind mount if we are in a user namespace having a parent namespace set,
+		// i.e. either one of uid base, shift, or the range is set, see user_namespaces(7).
+		if (uid_base != 0 || uid_shift != 0 || uid_range != -1) {
```

Comment:
Good catch, the documentation (http://man7.org/linux/man-pages/man7/user_namespaces.7.html) explicitly mentions the deliberate use of `-1`, according to http://code.metager.de/source/xref/gnu/glibc/bits/typesizes.h#30 it is an uint32, hence -1 will cast to maxuint.

I `#define`d an `UNMAPPED` constant inspired by http://lxr.free-electrons.com/source/kernel/sys.c#L473 to make it more clear.